os/arch/arm/src/s5j/sss/security_hal_wrapper.c

```diff
@@ -0,0 +1,1632 @@
+{
+	uint32_t i;
+	uint32_t j;
+
+	if (X->n == 0) {
+		return 0;
+	}
+
+	for (i = X->n - 1; i > 0; i--) {
+		if (X->p[i] != 0) {
+			break;
+		}
+	}
+	j = biL - hal_clz(X->p[i]);
+
+	return (i * biL) + j;
+}
+
+static uint32_t hal_mpi_size(const hal_mpi *X)
+{
+	return (hal_mpi_bitlen(X) + 7) >> 3;
+}
+
+static int hal_asn1_get_len(unsigned char **p, const unsigned char *end, uint32_t *len)
+{
+	if ((end - *p) < 1) {
+		return HAL_INVALID_ARGS;
+	}
+
+	if ((**p & 0x80) == 0) {
+		*len = *(*p)++;
+	} else {
+		switch (**p & 0x7F) {
+		case 1:
+			if ((end - *p) < 2) {
+				return HAL_INVALID_ARGS;
+			}
+			*len = (*p)[1];
+			(*p) += 2;
+			break;
+		case 2:
+			if ((end - *p) < 3) {
+				return HAL_INVALID_ARGS;
+			}
+			*len = ((uint32_t)(*p)[1] << 8) | (*p)[2];
+			(*p) += 3;
+			break;
+		case 3:
+			if ((end - *p) < 4) {
+				return HAL_INVALID_ARGS;
+			}
+			*len = ((uint32_t)(*p)[1] << 16) | ((uint32_t)(*p)[2] << 8) | (*p)[3];
+			(*p) += 4;
+			break;
+		case 4:
+			if ((end - *p) < 5) {
+				return HAL_INVALID_ARGS;
+			}
+			*len = ((uint32_t)(*p)[1] << 24) | ((uint32_t)(*p)[2] << 16) | ((uint32_t)(*p)[3] << 8) | (*p)[4];
+			(*p) += 5;
+			break;
+		default:
+			return HAL_INVALID_ARGS;
+		}
+	}
+
+	if (*len > (uint32_t)(end - *p)) {
+		return HAL_INVALID_ARGS;
+	}
+
+	return HAL_SUCCESS;
+}
+
+static int hal_asn1_get_tag(unsigned char **p, const unsigned char *end, uint32_t *len, int tag)
+{
+	if ((end - *p) < 1) {
+		return HAL_INVALID_ARGS;
+	}
+
+	if (**p != tag) {
+		return HAL_INVALID_ARGS;
+	}
+
+	(*p)++;
+
+	return hal_asn1_get_len(p, end, len);
+}
+
+static int hal_mpi_read_binary(hal_mpi *X, const unsigned char *buf, uint32_t buflen)
+{
+	int ret;
+	uint32_t i;
+	uint32_t j;
+	uint32_t const limbs = CHARS_TO_LIMBS(buflen);
+
+	/* Ensure that target MPI has exactly the necessary number of limbs */
+	if (X->n != limbs) {
+		hal_mpi_free(X);
+		hal_mpi_init(X);
+		HAL_MPI_CHK(hal_mpi_grow(X, limbs));
+	}
+
+	HAL_MPI_CHK(hal_mpi_lset(X, 0));
+
+	for (i = buflen, j = 0; i > 0; i--, j++) {
+		X->p[j / ciL] |= ((uint64_t)buf[i - 1]) << ((j % ciL) << 3);
+	}
+
+cleanup:
+	return ret;
+}
+
+static int hal_mpi_write_binary(const hal_mpi *X, unsigned char *buf, uint32_t buflen)
+{
+	uint32_t stored_bytes = X->n * ciL;
+	uint32_t bytes_to_copy;
+	unsigned char *p;
+	uint32_t i;
+
+	if (stored_bytes < buflen) {
+		bytes_to_copy = stored_bytes;
+		p = buf + buflen - stored_bytes;
+		memset(buf, 0, buflen - stored_bytes);
+	} else {
+		bytes_to_copy = buflen;
+		p = buf;
+		for (i = bytes_to_copy; i < stored_bytes; i++) {
+			if (GET_BYTE(X, i) != 0) {
+				return HAL_NOT_ENOUGH_MEMORY;
+			}
+		}
+	}
+
+	for (i = 0; i < bytes_to_copy; i++) {
+		p[bytes_to_copy - i - 1] = GET_BYTE(X, i);
+	}
+
+	return HAL_SUCCESS;
+}
+
+static int hal_asn1_write_tag(unsigned char **p, unsigned char *start, unsigned char tag)
+{
+	if (*p - start < 1) {
+		return -HAL_INVALID_ARGS;
+	}
+
+	*--(*p) = tag;
+
+	return 1;
```

Comment:
some functions return HAL_SUCCESS(c.f. hal_mpi_write_binary), here rutnrs 1;
Why don't you make return types first?
Same opinion for error case.